### ISSUE TITLE
Argument type is partially unknown after using `issubclass` with a generic class

### ISSUE THREAD

**Issue by ITProKyle** [microsoft/pyright#6716]
Not sure if this is a bug or an incorrect usage on my end. It has been working with all releases for the past 5 months until 1.1.339 so I'm opening it as a bug report.

**Describe the bug**

```
/Users/me/repos/project/test.py
  /Users/me/repos/project/test.py:58:13 - error: Argument type is partially unknown
    Argument corresponds to parameter "__value" in function "__setitem__"
    Argument type is "type[LookupHandler[ContextTypeVar@register_lookup_handler]] | type[LookupHandler[Unknown]]" (reportUnknownArgumentType)
1 error, 0 warnings, 0 informations 
```

The following code results in the above error where I would expect no error as it was prior to 1.1.339.

**Code or Screenshots**

```python
"""Test something."""
from __future__ import annotations

import logging
from typing import Any, Generic, TypeVar

LOGGER = logging.getLogger(__name__)


class Context0:
    """Example object."""


class Context1:
    """Example object."""


ContextTypeVar = TypeVar(
    "ContextTypeVar", "Context0", "Context1", "Context0 | Context1"
)
"""Type variable for context type."""


class LookupHandler(Generic[ContextTypeVar]):
    """Base class for lookup handlers."""

    @classmethod
    def handle(
        cls,
        __value: str,
        context: ContextTypeVar,
    ) -> Any:
        """Perform the lookup.

        Args:
            __value: Parameter(s) given to the lookup.
            context: The current context object.

        """
        raise NotImplementedError


LOOKUP_HANDLERS: dict[str, type[LookupHandler[Any]]] = {}


def register_lookup_handler(
    lookup_type: str, handler: type[LookupHandler[Any]] | type
) -> None:
    """Register a lookup handler.

    Args:
        lookup_type: Name to register the handler under
        handler: Lookup handler.

    """
    try:
        if issubclass(handler, LookupHandler):
            LOOKUP_HANDLERS[lookup_type] = handler
            return
    except Exception:  # noqa: BLE001
        LOGGER.debug("failed to validate lookup handler", exc_info=True)
    raise TypeError(
        f"lookup {handler} must be a subclass of test.LookupHandler"
    )
```

In the code above, `handler` is user provided. The `issubclass` is being performed to _fail fast_ instead of waiting to raise an error when attempting to use it.

**VS Code extension or command-line**

`npm exec --no -- pyright --venvpath ./ test.py` with pyright 1.1.339

**Comment by erictraut** [comment]
I don't consider this a bug. There were some recent changes regarding the use of a "bare" `type` in an annotation prompted by [this discussion in this typing forums](https://discuss.python.org/t/inconsistencies-between-type-and-type/37404/12). That explains why you saw this recent change.

The consensus in the typing community is that `type` should always be treated the same as `type[Unknown]`. When used with a `issubclass(handler, LookupHandler)` type guard, `type[Unknown]` is narrowed to `type[LookupHandler[Unknown]]`.

For comparison, mypy currently narrows the type to `type[LookupHandler[Any]] | type[Any]`.

Your function `register_lookup_handler` appears to require that `handler` is always a subclass of `LookupHandler`, so it's not clear to me why your static type annotation accepts `type`. Perhaps this is because you've simplified the code in the example?

In any case, I think pyright is working correctly here, so I'm going to close the issue.

**Comment by ITProKyle** [comment]
> Your function register_lookup_handler appears to require that handler is always a subclass of LookupHandler, so it's not clear to me why your static type annotation accepts type. Perhaps this is because you've simplified the code in the example?

Ya, it's simplified quite a bit to pull out other code that is not directly related to `issubclass(handler, LookupHandler)` becoming `type[LookupHandler[Unknown]] | type[LookupHandler[Any]]`

The unsimplified version takes in a parameter that is a string or subclass of `LookupHandler`. If its a string, it gets passed into another generalized function that wraps `importlib.import_module`. The portion that is user provided would be a string that we then import. The return type of that function is `type | Callable[..., Any]` since we can't know what it's actually going to be until we start inspecting it to make sure it is a subclass of the expected generic. I tested casting this return to `type[Any]` (prior to the `issubclass` check) with the same result.
